Bookstore: Use Redux in React components.

VicPeralta opened the pull request:
# Bookstore: Use Redux in React components

In this milestone:

- No linter errors.
- Correct GitFlow was used.
- Repo documented professionally.
- State kept immutable in reducers.
- Imported and used Redux `Provider`.
- Connected the App to the Redux Store.
- The list of displayed books comes from the Redux store, 2 books were added at the initial stage in order to present something.
- `useSelector `was used to consume the state.
- Add new book functionality was added using Redux `useDispatch`.
- Remove book functionality was added using Redux `useDispatch`.
- For the categories page, functionality was provided to the Check status button using Redux `useDispatch`.

VicPeralta committed:
Change createStore to configureStore

modified:   src/redux/configureStore.js

VicPeralta committed:
Add functionality for Books section using Redux

modified:   package.json
modified:   src/components/Book/Book.js
modified:   src/components/Books/Books.js
modified:   src/components/Input-Form/Input-Form.js
modified:   src/index.js
modified:   src/pages/books-page.js
deleted:    src/redux/books/books.js
deleted:    src/redux/books/books.test.js
new file:   src/redux/books/booksSlice.js
modified:   src/redux/configureStore.js

VicPeralta committed:
Add functionality for Categories section using Reduc

Add some style

modified:   src/components/Book/Book.css
modified:   src/components/Book/Book.js
modified:   src/components/Books/Books.css
modified:   src/components/Books/Books.js
modified:   src/components/Categories/Categories.js
modified:   src/components/Input-Form/Input-Form.css
modified:   src/components/Input-Form/Input-Form.js
modified:   src/components/NavBar/NavBar.css
modified:   src/index.css
new file:   src/pages/books-page.css
modified:   src/pages/books-page.js
new file:   src/pages/categories-page.css
modified:   src/pages/categories-page.js
new file:   src/redux/categories/categoriesSlice.js
modified:   src/redux/configureStore.js

VicPeralta committed:
Update readme file

Give some styling for mobile devices

modified:   desktop.png
modified:   mobile.png
modified:   src/components/Input-Form/Input-Form.css
modified:   src/components/NavBar/NavBar.css
modified:   tablet.png

Gambit142 commented on src/redux/books/booksSlice.js:
```diff
@@ -0,0 +1,43 @@
+  },
+];
+
+export const booksSlice = createSlice({
+  name: 'books',
+  initialState: {
+    books: initialState,
+  },
+  reducers: {
+    removeBook: (state, action) => {
+      const newBooks = state.books.filter((book) => (book.id !== action.payload.id));
+      return {
+        books: newBooks,
+      };
+    },
+    addBook: (state, action) => (
+      {
+        books: [
+          ...state.books,
+          {
+            id: action.payload.id,
+            title: action.payload.title,
+            author: action.payload.author,
+          },
+        ],
+      }),
+  },
+});
```
- [ ] It would have been nice if you left the action creators and the reducer the way they were. Don't get me wrong, using `createSlice` is not wrong, but since we are practising **_React-Redux_**, it would have been better to use pure functions as your action creators and reducers. One of the main goals of this project is to make you understand the concept of using react-redux. That is the reason why each milestone specifies what should be done. Kindly revert to the former way you created the action creators and reducers. 👍 

Kindly see requirement below:

![image](https://user-images.githubusercontent.com/77106962/160718171-d6b63951-e4bb-44bb-ae95-7c84cf9a578a.png)

Gambit142 commented on src/components/Input-Form/Input-Form.js:
```diff
@@ -1,17 +1,49 @@
+  const handler = (e) => {
+    e.preventDefault();
+    const title = document.getElementById('title');
+    const author = document.getElementById('author');
```
- [ ] You don't need to get the `title` and `author` elements this way nor do you need to change their status this way. Kindly use local state i.e React Hook `useState` instead of targeting them this way. You can read this [article](https://reactjs.org/docs/hooks-state.html) for a better understanding of the usage of `useState` as a react Hook.

Gambit142 commented on src/redux/categories/categoriesSlice.js:
```diff
@@ -0,0 +1,21 @@
+import { createSlice } from '@reduxjs/toolkit';
+
+export const categoriesSlice = createSlice({
+  name: 'categories',
+  initialState: {
+    status: 'Under construction',
+    show: false,
+  },
+  reducers: {
+    statusCategories: (state) => (state.status),
+    toggleShow: (state) => (
+      {
+        ...state,
+        show: !state.show,
+      }
+    ),
+  },
+});
+
+export const { statusCategories, toggleShow } = categoriesSlice.actions;
+export default categoriesSlice.reducer;
```
- [ ] Kindly use the action creators and reducer in the `categories.js` instead of this one as a slice file. 👍 

Kindly see the requirement below

![image](https://user-images.githubusercontent.com/77106962/160717972-772380b5-0f0b-4351-bd05-b57b976ba7dd.png)

Gambit142 reviewed (changes requested):
Hi @VicPeralta 👋, 

Good job so far!

## Highlights

- Linters checks passed ✔️ 
- Very professional `README.md` file ✔️ 

However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

VicPeralta committed:
Make changes requested by reviewer

Revert to use the old syntax instead of the new slice syntax, for the books and categories reducers.
In the form, instead of using the standard validations methods, it uses the state hook.

modified:   src/components/Book/Book.js
modified:   src/components/Categories/Categories.js
modified:   src/components/Input-Form/Input-Form.css
modified:   src/components/Input-Form/Input-Form.js
modified:   src/components/NavBar/NavBar.css
modified:   src/pages/books-page.js
modified:   src/pages/categories-page.css
new file:   src/redux/books/books.js
modified:   src/redux/categories/categories.js
deleted:    src/redux/categories/categories.test.js
modified:   src/redux/configureStore.js
new file:   src/redux/configureStoreSlices.js

omar-labana reviewed (approved):
# **Approved ✔️**

Hi @VicPeralta,

Congratulations on completing this project! 👏🏼 🎉

You have correctly implemented all of the requirements📃 correctly and without any errors⚠️ or bugs🪲.

Cheers and good luck in your next milestones! 🍀

Feel free to leave any questions or comments in the PR thread(don't forget to tag me 😄) if something is not 100% clear.

---
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
